go-controller/pkg/ovn/controller/egressfirewall/egressfirewall.go

```diff
@@ -72,75 +92,130 @@ type destination struct {
+	// Based on this flag we can omit clusterSubnet exclusion from the related ACL.
+	// For dns-based rules, EgressDNS won't add ips from clusterSubnet to the address set.
+	clusterSubnetIntersection bool
 }
 
-// newEgressFirewallRule creates a new egressFirewallRule. For the logging level, it will pick either of
-// aclLoggingAllow or aclLoggingDeny depending if this is an allow or deny rule.
-func (oc *DefaultNetworkController) newEgressFirewallRule(rawEgressFirewallRule egressfirewallapi.EgressFirewallRule, id int) (*egressFirewallRule, error) {
-	efr := &egressFirewallRule{
-		id:     id,
-		access: rawEgressFirewallRule.Type,
+type matchKind int
+
+type EFController struct {
+	sync.RWMutex // used only for unit testing
+	name         string
+	zone         string
+	ruleCounter  sync.Map
+
+	// libovsdb northbound client interface
+	nbClient libovsdbclient.Client
+	kube     *kube.KubeOVN
+
+	nodeLister      corelisters.NodeLister
+	namespaceLister corelisters.NamespaceLister
+	efLister        v2.EgressFirewallLister
+
+	controller     controller.Controller
+	nodeController controller.Controller
+	networkManager networkmanager.Interface
+	// dnsNameResolver is used for resolving the IP addresses of DNS names
+	// used in egress firewall rules
+	dnsNameResolver dnsnameresolver.DNSNameResolver
+	observManager   *observability.Manager
+}
+
+func NewEFController(
+	name string,
+	zone string,
+	kube *kube.KubeOVN,
+	nbClient libovsdbclient.Client,
+	namespaceLister corelisters.NamespaceLister,
+	nodeInformer coreinformers.NodeInformer,
+	efInformer v1.EgressFirewallInformer,
+	networkManager networkmanager.Interface,
+	dnsNameResolver dnsnameresolver.DNSNameResolver,
+	observManager *observability.Manager,
+) (*EFController, error) {
+	c := &EFController{
+		name:            name,
+		zone:            zone,
+		nbClient:        nbClient,
+		kube:            kube,
+		nodeLister:      nodeInformer.Lister(),
+		namespaceLister: namespaceLister,
+		efLister:        efInformer.Lister(),
+		networkManager:  networkManager,
+		dnsNameResolver: dnsNameResolver,
+		observManager:   observManager,
+		ruleCounter:     sync.Map{},
 	}
 
-	// Validate the egress firewall rule destination and update the appropriate
-	// fields of efr.
-	var err error
-	efr.to.cidrSelector, efr.to.dnsName, efr.to.clusterSubnetIntersection, efr.to.nodeSelector, err =
-		util.ValidateAndGetEgressFirewallDestination(rawEgressFirewallRule.To)
-	if err != nil {
-		return efr, err
+	controllerConfig := &controller.ControllerConfig[egressfirewallapi.EgressFirewall]{
+		RateLimiter: workqueue.DefaultTypedControllerRateLimiter[string](),
+		Informer:    efInformer.Informer(),
+		Lister:      efInformer.Lister().List,
+		MaxAttempts: controller.InfiniteAttempts,
+		Reconcile:   c.sync,
+		// Node changes can require reconcile, need to always reconcile. Should not be many updates to EF.
+		ObjNeedsUpdate: func(_, _ *egressfirewallapi.EgressFirewall) bool {
```

Comment:
this one should be changed to ` old == nil || new == nil || !reflect.DeepEqual(old.Spec, new.Spec)` to avoid a full reconcile on every node's status condition update